### PR TITLE
IGNITE-15133 Cursors and watches cleanup logic on node left.

### PR DESCRIPTION
**Pull request by sanpwc**

### CODE REVIEW

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java [hunk 1]
```diff
@@ -230,23 +232,33 @@ else if (clo.command() instanceof RangeCommand) {
...
 
                 clo.result(cursorId);
             }
             else if (clo.command() instanceof CursorNextCommand) {
                 CursorNextCommand cursorNextCmd = (CursorNextCommand) clo.command();
 
-                assert cursors.containsKey(cursorNextCmd.cursorId());
+                IgniteBiTuple<Cursor<?>, CursorMeta> cursorDesc = cursors.get(cursorNextCmd.cursorId());
+
+                if (cursorDesc == null) {
```
Comment (vldpyatkov) [inline modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:250]: Why did the cursor can not be found?
I think, need to add an extra information (the node left the cluster), also assertion for that.

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java [hunk 2]
```diff
@@ -74,7 +76,7 @@
     private final KeyValueStorage storage;
 
     /** Cursors map. */
-    private final Map<IgniteUuid, IgniteBiTuple<Cursor<?>, CursorType>> cursors;
+    private final Map<IgniteUuid, IgniteBiTuple<Cursor<?>, CursorMeta>> cursors;
```
Comment (vldpyatkov) [inline modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:79]: It looks complicated. If I were you, I would add Cursor to CursorMeta.

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java [hunk 3]
```diff
@@ -303,10 +321,38 @@ else if (clo.command() instanceof WatchExactKeysCommand) {
...
+                        new CursorMeta(CursorType.WATCH, watchCmd.requesterNodeId())
+                    )
+                );
 
                 clo.result(cursorId);
             }
+            else if (clo.command() instanceof CursorsCloseCommand) {
+                CursorsCloseCommand cursorsCloseCmd = (CursorsCloseCommand) clo.command();
+
+                for (Iterator<IgniteBiTuple<Cursor<?>, CursorMeta>> cursorsIter = cursors.values().iterator();
```
[the remark this answers is not in the document]
Comment (sanpwc) [reply modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:337]: Fixed.

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java (hunk 1 above)
Comment (sanpwc) [reply modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:250]: In real life you should never get here, cause it's only possible to get null if we try to retrieve next element for a node that have already left topology. However it's ease testing dramatically and we still respect the cursor contract.

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java (hunk 2 above)
Comment (sanpwc) [reply modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java]: Fixed.

File: modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java (hunk 1 above)
Comment (sanpwc) [reply modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/raft/MetaStorageListener.java:254]: Following comment was added:
```
                // In real life we should never get here, cause it's only possible to get null if we try to retrieve
                // next element for a cursor of a node that has already left topology.
                // However testing with such an approach is much easier and we still respect the cursor's contract.
```

**Pull request closed without merging**